server/src/main/java/org/apache/druid/indexing/overlord/IndexerMetadataStorageCoordinator.java

```diff
@@ -123,17 +123,30 @@ Collection<DataSegment> retrieveUsedSegmentsForIntervals(
       Segments visibility
   );
 
+  /**
+   * see {@link #retrieveUnusedSegmentsForInterval(String, Interval, Integer)}
+   */
+  List<DataSegment> retrieveUnusedSegmentsForInterval(String dataSource, Interval interval);
+
   /**
    * Retrieve all published segments which include ONLY data within the given interval and are marked as unused from the
    * metadata store.
    *
-   * @param dataSource The data source the segments belong to
-   * @param interval   Filter the data segments to ones that include data in this interval exclusively.
+   * @param dataSource  The data source the segments belong to
+   * @param interval    Filter the data segments to ones that include data in this interval exclusively.
+   * @param maxSegments The maximum number of unused segments to retreive. If null, no limit is applied.
    *
    * @return DataSegments which include ONLY data within the requested interval and are marked as unused. Segments NOT
    * returned here may include data in the interval
    */
-  List<DataSegment> retrieveUnusedSegmentsForInterval(String dataSource, Interval interval);
+  default List<DataSegment> retrieveUnusedSegmentsForInterval(
+      String dataSource,
+      Interval interval,
+      @Nullable Integer maxSegments
```

Comment:
This behaviour is a little confusing to me, since the `default` implementation does not fulfill the contract. Perhaps instead it should be reversed: a default for `retrieveUnusedSegmentsForInterval(String, Interval)`, can be provided, that calls to `retrieveUnusedSegmentsForInterval(String, Interval, null);`

Reply:
fixed